core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+    {
+        return fragmentDependency;
+    }
+
+    @VisibleForTesting
+    Set<StageExecution> getActiveSources()
+    {
+        return activeSources;
+    }
+
+    private void removeCompletedStages()
+    {
+        Set<StageExecution> completedStages = activeSources.stream()
+                .filter(this::isStageCompleted)
+                .collect(toImmutableSet());
+        // remove completed stages outside of Java stream to prevent concurrent modification
+        completedStages.forEach(this::removeCompletedStage);
+    }
+
+    private void removeCompletedStage(StageExecution stage)
+    {
+        // start all stages that depend on completed stage
+        PlanFragmentId fragmentId = stage.getFragment().getId();
+        fragmentDependency.outgoingEdgesOf(fragmentId).stream()
+                .map(FragmentsEdge::getTarget)
+                // filter stages that depend on completed stage only
+                .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+        fragmentDependency.removeVertex(fragmentId);
+        fragmentTopology.removeVertex(fragmentId);
+        activeSources.remove(stage);
+    }
+
+    private void unblockStages()
+    {
+        // find stages that are blocked on full task output buffer
+        Set<PlanFragmentId> blockedFragments = activeSources.stream()
+                .filter(StageExecution::isAnyTaskBlocked)
+                .map(stage -> stage.getFragment().getId())
+                .collect(toImmutableSet());
+        // start immediate downstream stages so that data can be consumed
+        blockedFragments.stream()
+                .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
+                .map(FragmentsEdge::getTarget)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    private void startStage(StageExecution stage)
+    {
+        if (fragmentDependency.outDegreeOf(stage.getFragment().getId()) > 0) {
+            // if there are any dependent stages then reschedule when stage is completed
+            stage.addStateChangeListener(state -> {
+                if (isStageCompleted(stage)) {
+                    notifyReschedule();
+                }
+            });
+        }
+        activeSources.add(stage);
+    }
+
+    private void notifyReschedule()
+    {
+        SettableFuture<Void> rescheduleFuture;
+        synchronized (this) {
+            rescheduleFuture = this.rescheduleFuture;
+            this.rescheduleFuture = SettableFuture.create();
+        }
+        // notify listeners outside of the critical section
+        rescheduleFuture.set(null);
+    }
+
+    private boolean isStageCompleted(StageExecution stage)
+    {
+        State state = stage.getState();
+        return state == SCHEDULED || state == RUNNING || state == FLUSHING || state.isDone();
```

Comment:
note: this allows for scheduling of dependent stage before parent stage completes or enters flushing state.
I've benchmarked code where we consider `flushing and done` stages to be completed, but results were not better (there was even a slight duration regression for tpcds):



label | TPCH wall time | TPC-DS wall time | TPCH CPU time | TPC-DS CPU time | TPCH peak mem | TPC-DS peak mem
-- | -- | -- | -- | -- | -- | --
before staged unpart | 769.708500 | 1925.400000 | 98571.0 | 237859.442667 | 2.093233e+09 | 1.147734e+09
after staged unpart | 758.691167 | 1816.289500 | 99912.3 | 242806.413833 | 2.082893e+09 | 1.132478e+09
after staged unpart compl | 757.288833 | 1848.024833 | 99515.5 | 242440.454833 | 2.082462e+09 | 1.101463e+09



label | TPCH wall time | TPC-DS wall time | TPCH CPU time | TPC-DS CPU time | TPCH peak mem | TPC-DS peak mem
-- | -- | -- | -- | -- | -- | --
before staged part | 937.515833 | 1184.331667 | 111597.4 | 135037.311333 | 2.144486e+09 | 1.284780e+09
after staged part | 829.780167 | 1133.327333 | 111747.4 | 132606.885667 | 2.141816e+09 | 1.267524e+09
after staged part compl | 827.721667 | 1153.291667 | 110661.2 | 133056.365000 | 2.148532e+09 | 1.246323e+09

Reply:
This is useful comment - please put that in the code.